File: C#/16-Ordenação de Vetores/ShellSort.cs
using System;

class ShellSort
{
    //Método shell sort para ordenação de vetores
    public static int[] shellSort(int[] vetor)
    {
        int h = 1;
        int n = vetor.Length;

        while(h < n)
        {
            h = h * 3 + 1;
        }

        h = h / 3;
        int c;
        int j;

        while(h > 0)
        {
            for (int i = h; i < n; i++) 
            {
                c = vetor[i];
                j = i;
                while (j >= h && vetor[j - h] > c) 
                {
                    vetor[j] = vetor[j - h];
                    j = j - h;
                }

                vetor[j] = c;
            }

            h = h / 2;
        }
        return vetor;
    }

    static void Main(string[] args)
    {
        int[] numeros = new int[] { 2, 5, 1, 3, 4 };
        Console.WriteLine("Vetor original:");
        foreach (int numero in numeros)
        {
            Console.WriteLine(numero);
        }

        Console.WriteLine("Pós shell sort");

        //Vamos invocar o método passando vetor numeros como argumento
        shellSort(numeros);
        //Exibindo vetor ordenado
        foreach (int numero in numeros)
        {
            Console.WriteLine(numero);
        }

        Console.ReadLine();
    }
}

File: JavaScript/13-Ordenação de Vetores/ShellSort.js
function shellSort(vetor) {

    var h = vetor.length / 2;

    while (h > 0) {

        for (i = h; i < vetor.length; i++) {

            var j = i;
            var c = vetor[i];

            while (j >= h && vetor[j - h] > c) {
                vetor[j] = vetor[j - h];
                j = j - h;
            }

            vetor[j] = c;

        }

        if (h == 2) {
            h = 1;
        } else {
            h = parseInt(h * 5 / 11);
        }
        
    }

    return vetor;

}

var vet = [4, 2, 1, 8, 3];

document.write("Vetor original" + "<br/>");

//Exibindo vetor original
for (var i = 0; i < vet.length; i++) {
    document.write(vet[i] + " ");
}

document.write("<br/>");

//Ordenando com shellsort

document.write("Pós ShellSort" + "<br/>");
var ord = shellSort(vet);

for (var i = 0; i < ord.length; i++) {
    document.write(ord[i] + " ");
}